jruby secure download

cpuguy83 opened the pull request:
JRuby download was recently updated with to use a proper https URL, updated here.

yosifkit commented:
Build test of #391; 592994d9ebd00166fa07ec34f4ade9546098e04d (`jruby`):

```console
$ ./bashbrew/bashbrew.sh build https://raw.githubusercontent.com/docker-library/official-images/592994d9ebd00166fa07ec34f4ade9546098e04d/library/jruby
Fetching jruby (git://github.com/cpuguy83/docker-jruby) ...
Processing jruby:1.7 ...
Processing jruby:1.7.18 ...
Processing jruby:latest ...
Processing jruby:1.7-onbuild ...
Processing jruby:1.7.18-onbuild ...
```

yosifkit commented:
LGTM

cpuguy83 commented:
Really, it built for you?
Having issues with java:8 CA certs here, so it can't `gem install bundler`

yosifkit commented:
Here is my build log:

```console
$ docker build -t jruby:1.7 stackbrew/bashbrew/src/github.com/cpuguy83/docker-jruby/1.7
Sending build context to Docker daemon 4.096 kB
Sending build context to Docker daemon 
Step 0 : FROM java:8
 ---> aec8201c9d63
Step 1 : ENV JRUBY_VERSION 1.7.18
 ---> Running in 9f0011aeb019
 ---> c685a5368d73
Removing intermediate container 9f0011aeb019
Step 2 : RUN mkdir /opt/jruby   && curl https://s3.amazonaws.com/jruby.org/downloads/1.7.18/jruby-bin-${JRUBY_VERSION}.tar.gz   | tar -zxC /opt/jruby --strip-components=1   && update-alternatives --install /usr/local/bin/ruby ruby /opt/jruby/bin/jruby 1
 ---> Running in ae4ff17640b3
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 31.1M  100 31.1M    0     0  2607k      0  0:00:12  0:00:12 --:--:-- 3440k
update-alternatives: using /opt/jruby/bin/jruby to provide /usr/local/bin/ruby (ruby) in auto mode
 ---> a1a46a4257d3
Removing intermediate container ae4ff17640b3
Step 3 : ENV PATH /opt/jruby/bin:$PATH
 ---> Running in 4dc6232d07dc
 ---> cb45d6ab25fd
Removing intermediate container 4dc6232d07dc
Step 4 : RUN echo 'gem: --no-rdoc --no-ri' >> /.gemrc
 ---> Running in f84d0ca88325
 ---> 536469bbfd59
Removing intermediate container f84d0ca88325
Step 5 : ENV GEM_HOME /usr/local/bundle
 ---> Running in 19dcef8c53cc
 ---> 7438d4a3ba29
Removing intermediate container 19dcef8c53cc
Step 6 : ENV PATH $GEM_HOME/bin:$PATH
 ---> Running in 6f8390fd15eb
 ---> f24924df7121
Removing intermediate container 6f8390fd15eb
Step 7 : RUN gem install bundler 	&& bundle config --global path "$GEM_HOME" 	&& bundle config --global bin "$GEM_HOME/bin"
 ---> Running in 8e14acef6c05
Successfully installed bundler-1.7.11
1 gem installed
 ---> a2dba21c2cec
Removing intermediate container 8e14acef6c05
Step 8 : ENV BUNDLE_APP_CONFIG $GEM_HOME
 ---> Running in 780bb0e53960
 ---> 28f2d2d8e23e
Removing intermediate container 780bb0e53960
Step 9 : CMD irb
 ---> Running in c2ff3600834b
 ---> ea77a06c0d2e
Removing intermediate container c2ff3600834b
Successfully built ea77a06c0d2e
```

And here is my test using the image id just built above:

```console
$ docker run -it --rm ea77a06c0d2e gem install bundler
Successfully installed bundler-1.7.11
1 gem installed
```

tianon commented:
I just tested too:
```console
$ docker build 1.7
Sending build context to Docker daemon 4.096 kB
Sending build context to Docker daemon 
Step 0 : FROM java:8
 ---> aec8201c9d63
Step 1 : ENV JRUBY_VERSION 1.7.18
 ---> Running in 913edc671c28
 ---> 21eb6cf8b917
Removing intermediate container 913edc671c28
Step 2 : RUN mkdir /opt/jruby   && curl https://s3.amazonaws.com/jruby.org/downloads/1.7.18/jruby-bin-${JRUBY_VERSION}.tar.gz   | tar -z+ docker build -t jruby:1.7 /home/tianon/docker/stackbrew/bashbrew/src/github.com/cpuguy83/docker-jruby/1.7
Sending build context to Docker daemon 4.096 kB
Sending build context to Docker daemon 
Step 0 : FROM java:8
 ---> aec8201c9d63
Step 1 : ENV JRUBY_VERSION 1.7.18
 ---> Running in 913edc671c28
 ---> 21eb6cf8b917
Removing intermediate container 913edc671c28
Step 2 : RUN mkdir /opt/jruby   && curl https://s3.amazonaws.com/jruby.org/downloads/1.7.18/jruby-bin-${JRUBY_VERSION}.tar.gz   | tar -zxC /opt/jruby --strip-components=1   && update-alternatives --install /usr/local/bin/ruby ruby /opt/jruby/bin/jruby 1
 ---> Running in 9a86f3d20754
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 31.1M  100 31.1M    0     0  2432k      0  0:00:13  0:00:13 --:--:-- 3230k
update-alternatives: using /opt/jruby/bin/jruby to provide /usr/local/bin/ruby (ruby) in auto mode
 ---> 05fc2753a4c9
Removing intermediate container 9a86f3d20754
Step 3 : ENV PATH /opt/jruby/bin:$PATH
 ---> Running in 7f933f32cecd
 ---> 001e15f295de
Removing intermediate container 7f933f32cecd
Step 4 : RUN echo 'gem: --no-rdoc --no-ri' >> /.gemrc
 ---> Running in 0980a14b6c8b
 ---> 2f2081773590
Removing intermediate container 0980a14b6c8b
Step 5 : ENV GEM_HOME /usr/local/bundle
 ---> Running in 4fb30fdbd2c5
 ---> f342f56be6cd
Removing intermediate container 4fb30fdbd2c5
Step 6 : ENV PATH $GEM_HOME/bin:$PATH
 ---> Running in eea0148f9f85
 ---> 868c06f92fd0
Removing intermediate container eea0148f9f85
Step 7 : RUN gem install bundler 	&& bundle config --global path "$GEM_HOME" 	&& bundle config --global bin "$GEM_HOME/bin"
 ---> Running in d74798606957
ERROR:  Could not find a valid gem 'bundler' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - certificate verify failed (https://rubygems.global.ssl.fastly.net/quick/Marshal.4.8/bundler-1.7.12.gemspec.rz)
ERROR:  Possible alternatives: bundler
time="2015-01-12T15:46:26-07:00" level="info" msg="The command [/bin/sh -c gem install bundler \t&& bundle config --global path \"$GEM_HOME\" \t&& bundle config --global bin \"$GEM_HOME/bin\"] returned a non-zero code: 2" 
xC /opt/jruby --strip-components=1   && update-alternatives --install /usr/local/bin/ruby ruby /opt/jruby/bin/jruby 1
 ---> Running in 9a86f3d20754
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 31.1M  100 31.1M    0     0  2432k      0  0:00:13  0:00:13 --:--:-- 3230k
update-alternatives: using /opt/jruby/bin/jruby to provide /usr/local/bin/ruby (ruby) in auto mode
 ---> 05fc2753a4c9
Removing intermediate container 9a86f3d20754
Step 3 : ENV PATH /opt/jruby/bin:$PATH
 ---> Running in 7f933f32cecd
 ---> 001e15f295de
Removing intermediate container 7f933f32cecd
Step 4 : RUN echo 'gem: --no-rdoc --no-ri' >> /.gemrc
 ---> Running in 0980a14b6c8b
 ---> 2f2081773590
Removing intermediate container 0980a14b6c8b
Step 5 : ENV GEM_HOME /usr/local/bundle
 ---> Running in 4fb30fdbd2c5
 ---> f342f56be6cd
Removing intermediate container 4fb30fdbd2c5
Step 6 : ENV PATH $GEM_HOME/bin:$PATH
 ---> Running in eea0148f9f85
 ---> 868c06f92fd0
Removing intermediate container eea0148f9f85
Step 7 : RUN gem install bundler 	&& bundle config --global path "$GEM_HOME" 	&& bundle config --global bin "$GEM_HOME/bin"
 ---> Running in d74798606957
ERROR:  Could not find a valid gem 'bundler' (>= 0), here is why:
          Unable to download data from https://rubygems.org/ - certificate verify failed (https://rubygems.global.ssl.fastly.net/quick/Marshal.4.8/bundler-1.7.12.gemspec.rz)
ERROR:  Possible alternatives: bundler
time="2015-01-12T15:46:26-07:00" level="info" msg="The command [/bin/sh -c gem install bundler \t&& bundle config --global path \"$GEM_HOME\" \t&& bundle config --global bin \"$GEM_HOME/bin\"] returned a non-zero code: 2" 
```

md5 commented:
Perhaps the failure is geo-DNS related? Looks like rubygems.global.ssl.fastly.net resolves to a number of different IP addresses depending on where you are in the world. Perhaps one of those IP addresses is using a bad cert.

cf. https://www.whatsmydns.net/#A/rubygems.global.ssl.fastly.net

cpuguy83 commented:
@md5 It can't be.
100% of the time I have an issue building with the 1.7 bin tarball on the java:8 image.
100% of the time it works on the java:7 image
100% of the time if I build from a nightly jruby snapshot it works.

But this has always been based off the java:8 image.

md5 commented:
@cpuguy83 what image id do you have for `java:8`? Looks like @yosifkit and @tianon have `aec8201c9d63`. That's what I got when I just pulled as well.

cpuguy83 commented:
@md5 That's the one.

yosifkit commented:
I just ran a `--no-cache` to make sure it was a full rebuild, and now it fails to build for me too. So, I am not sure why it worked the first time because that was a full build too as seen in my log.

md5 commented:
Looks like the certificate verification for JRuby is here: https://github.com/jruby/jruby-openssl/blob/master/src/main/java/org/jruby/ext/openssl/x509store/StoreContext.java#L636

Since `jruby-openssl` isn't doing any debug logging in this code, it looks like the only way to trace why it's failing is to step through in a debugger...

md5 commented:
I just found this helpful script: https://github.com/mislav/ssl-tools/blob/8b3dec4/doctor.rb

When I ran it against the same base layer that the `gem install bundler` is trying to run in (`e7dfacda306b` on my system), I see this:

```
$ docker run --rm e7dfacda306b sh -c 'curl -sSL https://raw.githubusercontent.com/mislav/ssl-tools/8b3dec4bedcc725a142fa9bc297610f8d09f5d9d/doc
tor.rb | ruby'
/opt/jruby/bin/jruby (1.9.3)
jruby-ossl 0.9.5: /usr/lib/ssl
SSL_CERT_DIR=""
SSL_CERT_FILE=""

HEAD https://status.github.com:443
OpenSSL::SSL::SSLError: certificate verify failed

The server presented a certificate that could not be verified:
  subject: /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert SHA2 High Assurance Server CA
  issuer: /C=US/O=DigiCert Inc/OU=www.digicert.com/CN=DigiCert High Assurance EV Root CA
  error code 20: unable to get local issuer certificate

Possible causes:
  `/usr/lib/ssl/cert.pem' does not exist
```

I couldn't find any good documentation about how one is supposed to configure a working set of trust roots for JRuby, but it looks like setting `SSL_CERT_DIR=/etc/ssl/certs` to make it use the certs from `ca-certificates` could work. Not sure if that's considered a best practice in the JRuby world, but here's what it does for me:

```
$ docker run -it --rm e7dfacda306b env SSL_CERT_DIR=/etc/ssl/certs gem install bundler
Fetching: bundler-1.7.12.gem (100%)
Successfully installed bundler-1.7.12
1 gem installed
```

md5 commented:
This discussion looks relevant: http://jira.codehaus.org/browse/JRUBY-5010?focusedCommentId=232209&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-232209

md5 commented:
It looks like the difference between `java:7` and `java:8` is that `cacerts` in `java:8` is a dangling symlink:

```
$ docker run --rm java:7 find / -name cacerts -ls
   188    4 -rw-------   1 root     root          384 Mar 25  2014 /etc/default/cacerts
   957  204 -rw-r--r--   1 root     root       206373 Jan  1 07:13 /etc/ssl/certs/java/cacerts
  3835    0 lrwxrwxrwx   1 root     root           27 Nov 13 14:24 /usr/lib/jvm/java-7-openjdk-amd64/jre/lib/security/cacerts -> /etc/ssl/certs/java/cacerts
$ docker run --rm java:8 find / -name cacerts -ls
   189    4 -rw-------   1 root     root          384 Mar 25  2014 /etc/default/cacerts
  3842    0 lrwxrwxrwx   1 root     root           27 Oct 14 11:30 /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -> /etc/ssl/certs/java/cacerts
```

md5 commented:
This looks like it could be related? https://github.com/docker-library/java/issues/19

cpuguy83 commented:
Wow @md5, thanks for looking at this!

cpuguy83 commented:
Ok, I've updated this.

I am now using `java:8-jre` as the main parent image since this is going to be enough for most people out there, and is significantly smaller.
Notable that the jre image does not have this SSL issue.

Added another section of images which includes the jdk.  So "1.7", "1.7.18", and "latest" all use jre, add "-jdk" to get the jdk versions, or add "-jre" to ensure jre versions.
jdk verion includes the workaround for the ssl issue, which is to `RUN /var/lib/dpkg/info/ca-certificates-java.postinst configure` before trying to install bundler`.

cpuguy83 commented:
Also updated with jruby 9000 prerelease build.

yosifkit commented:
Build test of #391; 965a96a19b66d2929b910339e73612b5bcf1b215 (`jruby`):

```console
$ ./bashbrew/bashbrew.sh build https://raw.githubusercontent.com/docker-library/official-images/965a96a19b66d2929b910339e73612b5bcf1b215/library/jruby
Fetching jruby (git://github.com/cpuguy83/docker-jruby) ...
Processing jruby:1.7 ...
Processing jruby:1.7.18 ...
Processing jruby:latest ...
Processing jruby:1.7-jre ...
Processing jruby:1.7.18-jre ...
Processing jruby:1.7-jdk ...
Processing jruby:1.7.18-jdk ...
Processing jruby:1.7-onbuild ...
Processing jruby:1.7.18-onbuild ...
Processing jruby:9.0.0.0 ...
Processing jruby:9.0.0.0-jre ...
Processing jruby:9.0.0.0.pre1 ...
Processing jruby:9.0.0.0.pre1-jre ...
Processing jruby:9.0.0.0-jdk ...
Processing jruby:9.0.0.0.pre1-jdk ...
Processing jruby:9.0.0.0-onbuild ...
Processing jruby:9.0.0.0.pre1-onbuild ...
```

tianon commented:
Fair enough for now; LGTM
